Add support for Sparse HLL PFADD

azuredream opened the pull request:
#2679

azuredream commented on src/redis/hyperloglog.h:
```diff
@@ -23,10 +23,16 @@ struct HllBufferPtr {
 };
 
+//createSizeTConfig("hll-sparse-max-bytes", NULL, MODIFIABLE_CONFIG, 0, LONG_MAX, server.hll_sparse_max_bytes, 3000, MEMORY_CONFIG, NULL, NULL),
+extern size_t HLL_SPARSE_MAX_BYTES;
```
Todo: Make this configurable.
Could anyone give me some suggestion on how to add a configuration?

chakaz replied:
I think that it's fine to keep as a define, at least for now.
If we'll find that users want to customize this, we'll add some flag or other means to set it.

azuredream commented on src/server/hll_family.cc:
```diff
@@ -75,26 +75,30 @@ OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values
-    int added = pfadd(StringToHllPtr(hll), (unsigned char*)value.data(), value.size());
+    // pfadd might resize the input string. A referance to string.data() makes it unnecessarily
+    // complicated
+    int added = pfadd(hll_ptr, (unsigned char*)value.data(), value.size());
```
Spent lots of time here. I finally decided not to use HllPtr because Sparse operation expands and reallocates string data. It's not easy to correctly update std::string object  in C.

A sds object is simple and fits the following logic well. 
Some overhead: Copying string to SDS, a reallocating on each sparse insertion, copy SDS back to string.

chakaz replied:
I agree. Indeed sds strings don't play nicely with Dragonfly strings, as they have incompatible layouts.
It looks like your current solution allocates for each additional item in the HLL. Can we make it so that it will only allocate for each additional item in an HLL *if the HLL is a sparse set*? To my understanding, once it becomes dense, there shouldn't be a need to do any further reallocations, because it is of a fixed size. So that could be a nice optimization (and a good trade-off, because the allocations will be on small sizes anyway)
What do you think?

azuredream commented on src/server/hll_family.cc:
```diff
@@ -75,26 +75,30 @@ OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values
   } else if (res.it->second.ObjType() != OBJ_STRING) {
     return OpStatus::WRONG_TYPE;
   } else {
     res.it->second.GetString(&hll);
```
If I understood correctly, GetString() returns a copy of data rather than reference, like robj in Redis.
This would cause some overhead to copy from and paste to string `hll`.
Since string cannot be used in C, more overhead to copy `hll_sds` from and paste to string `hll`.

chakaz replied:
Indeed. Please see my comment below.

azuredream commented:
It's working as expected now. Please let me know if you have any advice of the implementation. Not sure if I can use sds in .cc, maybe there is better practice.
Todo:
1. Add more test cases.
2. Code cleanup.

azuredream commented:
Todo:
3. A test case is failing intermittently. Please let me know if you have any advice...
```
[ RUN      ] HllFamilyTest.AddInvalid
/home/zhaoz12/git/dragonflymy/dragonfly/src/server/hll_family_test.cc:50: Failure
Value of: Run({"pfadd", "key", "1"})
Expected: is Key is not a valid HyperLogLog string value.
  Actual: i1 (of type facade::RespExpr), 
Wrong type: int
[  FAILED  ] HllFamilyTest.AddInvalid (17 ms)
```

chakaz commented on src/redis/hyperloglog.c:
Can you kindly specify which functions were *not* just copy pasted from Redis' code? I.e. which modifications you made?
Also, can you specify which version of Redis you took the code from? I want to make sure it was copied from a BSD licensed Redis and not from the new licensed code.
Thanks!

chakaz replied:
Just making sure you noticed this as well :)

chakaz commented:
@azuredream thank you, again, for your contribution!

chakaz commented on src/server/hll_family_test.cc:
```diff
@@ -26,6 +29,20 @@ TEST_F(HllFamilyTest, Simple) {
+  // (counting unique elements) of massive datasets (up to 2^64 elements). Its characteristics
+  // include high speed and low memory footprint (12KB). However, its computations come with a
+  // margin of error, with a standard error rate of 0.81%. got 0.83% error rate when unique_values =
+  // 200000 EXPECT_LT(std::abs(Run({"pfcount", "key"}) - unique_values)/unique_values, 0.0081);
```
```suggestion
  // 200000
  EXPECT_LT(std::abs(Run({"pfcount", "key"}) - unique_values)/unique_values, 0.1);
```

This should be high enough so that it wouldn't fail in practice, but will have some test value, no?

chakaz reviewed (commented):
Thanks again!
Can you kindly ping me when you're ready for the next iteration? I'm not sure if you wanted me to take a look yet or not :)

azuredream commented:
Sorry for the delay.
I think the code is ready to be reviewed. 

We also need to verify that the new change addresses the memory issue in #2679. Please let me know if you have any advice.

chakaz commented on src/server/hll_family_test.cc:
```diff
@@ -26,6 +29,20 @@ TEST_F(HllFamilyTest, Simple) {
+  int unique_values = 20000;
+  for (int i = 0; i < unique_values; ++i) {
+    std::string newkey = GenerateUniqueValue(i);
+    // sparse hll is promoted to dense at the 1660th+- insertion. But how to verify that?
```
How about using the size of the string?
Make sure that it grows until it reaches DENSE_SIZE and then continues in that size until the end of the test.

chakaz commented on src/server/hll_family_test.cc:
```diff
@@ -26,6 +29,20 @@ TEST_F(HllFamilyTest, Simple) {
+  for (int i = 0; i < unique_values; ++i) {
+    std::string newkey = GenerateUniqueValue(i);
+    // sparse hll is promoted to dense at the 1660th+- insertion. But how to verify that?
+    Run({"pfadd", "key", newkey});
```
It could be great to test adding multiple values at once (like the bug that I think I found). Can you randomize the number of entries to add, say between 1 and 20?
You could then do `i += count` instead of `++i` in the `for` loop to keep the size constant

chakaz reviewed (commented):
Thanks again for your contribution! I think we're very close, please see my comments, especially the one about the source license and which functions are yours and which were copied from Redis OSS

azuredream commented:
Updated.
1. Please let me know if want to keep sds operations in spares path only. We'll need to update string obj after each loop.
2. I think there is not a double free problem. Please correct me if I was wrong.
3. Added a random insertion test case as you suggested.
4. Do we need to check line by line of the code for the license concern? I'm sure there is not an identical function in this PR.

chakaz commented on src/server/hll_family.cc:
```diff
@@ -75,26 +75,52 @@ OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values
     updated += added;
   }
 
+  if (is_sparse) {
+    hll = string{hll_sds, sdslen(hll_sds)};
+    sdsfree(hll_sds);
+  }
```
You're right, I was mistaken here. Please ignore.

chakaz commented on src/server/hll_family_test.cc:
```diff
@@ -45,6 +71,41 @@ TEST_F(HllFamilyTest, MultipleValues) {
+  // cumulated pfadd result
+  for (int i = 0; i < insertions; ++i) {
+    // Number of values to insert
+    int numValues = dis(gen);
```
```suggestion
    int num_values = dis(gen);
```

chakaz commented on src/server/hll_family_test.cc:
```diff
@@ -26,6 +29,29 @@ TEST_F(HllFamilyTest, Simple) {
+  // This value varies if any parameter in hyperloglog.c changes.
+  int promote_i = 1660;
+  // Keep consistent with hyperloglog.c
+  int HLL_SPARSE_MAX_BYTES = 3000;
+  int HLL_DENSE_SIZE = 12304;
```
```suggestion
  int kHllSparseMaxBytes = 3000;
  int kHllDenseSize = 12304;
```

chakaz reviewed (commented):
Please see my small nits, and I apologize for my mistake before!

azuredream commented:
Updated.

1. Please see my reply under your comment at line 94.
4. The last commit of Hyperloglog.c in my Redis repo is on 11.26.2023. I it's fine?

chakaz commented on src/server/hll_family.cc:
```diff
@@ -75,26 +75,54 @@ OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values
+    int added;
+    if (is_sparse) {
+      // Inserting to sparse hll might extend it.
+      // Referring to string.data() makes it unnecessarily complicated
+      sds* hll_ptr = &hll_sds;
+      // 1 if sparse hll is promoted to dense
+      int promoted = 0;
+      added = pfadd_sparse(hll_ptr, (unsigned char*)value.data(), value.size(), &promoted);
+      hll_sds = *hll_ptr;
```
```suggestion
      // We can't use std::string with sds
      // `promoted` will be assigned 1 if sparse hll was promoted to dense
      int promoted = 0;
      added = pfadd_sparse(&hll_sds, (unsigned char*)value.data(), value.size(), &promoted);
```

chakaz reviewed (approved):
Woohoo, thanks a lot! :partying_face:

romange commented:
Thanks again @azuredream !
